core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala

```diff
@@ -370,38 +391,40 @@ object Checkpoints extends DeltaLogging {
           writeAction()
         }
         Iterator(writtenPath)
-      }.collect().head
+      }.collect()
 
     if (useRename) {
-      val src = new Path(writtenPath)
-      val dest = new Path(path)
-      val fs = dest.getFileSystem(hadoopConf)
-      var renameDone = false
-      try {
-        if (fs.rename(src, dest)) {
-          renameDone = true
-        } else {
-          // There should be only one writer writing the checkpoint file, so there must be
-          // something wrong here.
-          throw DeltaErrors.failOnCheckpoint(src, dest)
-        }
-      } finally {
-        if (!renameDone) {
-          fs.delete(src, false)
+      writtenPaths.zipWithIndex.foreach { case (writtenPath, index) =>
+        val src = new Path(writtenPath)
+        val dest = new Path(paths(index))
+        val fs = dest.getFileSystem(hadoopConf)
+        var renameDone = false
+        try {
+          if (fs.rename(src, dest)) {
+            renameDone = true
+          } else {
+            // There should be only one writer writing the checkpoint file, so there must be
+            // something wrong here.
+            throw DeltaErrors.failOnCheckpoint(src, dest)
+          }
+        } finally {
+          if (!renameDone) {
```

Comment:
do we need to delete all the part files when an exception occurs?

Reply:
yeah good call, let me think about that, probably just map the rename success and throw based on that instead

Reply:
Hmm actually begs a bigger question. So in the case renames aren't used, we're not cleaning anything up. In non-multipart, that's theoretically fine because if it fails there's nothing to clean up. Should that attempt to clean itself up? And for rename case, should it try all renames, and then if any of them fail try to delete all existing files (both that succeeded and failed to rename)?

Reply:
It's better to clean up temp paths if we are using `rename`. These temp files won't be cleaned up by our automatic log cleanup.

Reply:
So just cleanup all temp files for rename before throwing an exception and that's enough?

Reply:
Updated so it waits to try to rename all the files (and cleanup failures) before throwing the exception